master/internal/hpimportance/hpimportance.go

```diff
@@ -1,15 +1,242 @@
+	"os/exec"
+	"path"
+	"sort"
+	"strconv"
+	"strings"
+
+	log "github.com/sirupsen/logrus"
+
 	"github.com/determined-ai/determined/master/pkg/model"
 )
 
-func computeHPImportance(data []model.HPImportanceTrialData) map[string]float64 {
-	// This is a place-holder for a future Random Forest implementation to actually compute the
-	// importance of each hyperparameter.
-	output := make(map[string]float64)
-	for key := range data[0].Hparams {
-		output[key] = 0.0
+const (
+	maxDiffCompBatch = 4
+	minNumberTrials  = 50
+	idealBatchDiff   = 2
+	nReplications    = 2
+
+	arffFile       = "data.arff"
+	importanceFile = "importance.txt"
+)
+
+// The data needs to be put into an arff format.
+// Arff files require variable declaration at the top,
+// where order does matter. We need to keep define the Hps and
+// keep track of the order so the data columns will match.
+func createDataFile(data map[int][]model.HPImportanceTrialData,
+	experimentConfig *model.ExperimentConfig, dataFile string) (int, error) {
+	f, err := os.Create(dataFile)
+	if err != nil {
+		return 0, err
+	}
+	defer func() {
+		err = f.Close()
+		if err != nil {
+			log.WithError(err).Error("failed to close arff file")
+		}
+	}()
+
+	var arff strings.Builder
+
+	// create top of file based on exp config
+	_, err = arff.WriteString("@relation data\n\n@attribute metric numeric\n")
+	if err != nil {
+		return 0, err
+	}
+	var hpsOrder []string // HPs must be in the same order for the arff file
+	hps := experimentConfig.Hyperparameters
+
+	for key, element := range hps {
+		var st string
+		switch {
+		case element.ConstHyperparameter != nil:
+			continue
+		case element.CategoricalHyperparameter != nil:
+			hpsOrder = append(hpsOrder, key)
+
+			var values string
+			for _, catVal := range element.CategoricalHyperparameter.Vals {
+				values += fmt.Sprintf("%v,", catVal)
+			}
+			st = fmt.Sprintf("@attribute %s {%s}\n", key, values)
+		default:
+			hpsOrder = append(hpsOrder, key)
+			st = fmt.Sprintf("@attribute %s numeric\n", key)
+		}
+		_, err = arff.WriteString(st)
+		if err != nil {
+			return 0, err
+		}
+	}
+	_, err = arff.WriteString("@attribute numBatches numeric\n\n")
+	if err != nil {
+		return 0, err
+	}
+
+	// Now we have to add the data to the file
+	_, err = arff.WriteString("\n@data\n")
+	if err != nil {
+		return 0, err
+	}
+
+	// First, get batch ids unless better way to get them in go?
+	var batches []int
+	for k := range data {
+		batches = append(batches, k)
+	}
+	sort.Sort(sort.Reverse(sort.IntSlice(batches)))
+
+	totalNumTrials := 0
+	maxNumBatches := batches[0]
+	for _, batchID := range batches {
+		if batchID < maxNumBatches/maxDiffCompBatch {
+			break
+		}
+		if totalNumTrials > minNumberTrials && batchID <= maxNumBatches/idealBatchDiff {
+			break
+		}
+		batchIDStr := fmt.Sprintf("%v", batchID)
+		for _, trial := range data[batchID] {
+			var st string
+			st += fmt.Sprintf("%v", trial.Metric)
+			hparamsVals := trial.Hparams
+			for _, hp := range hpsOrder {
```

Comment:
`trial.Hparams` is a `hparamSample`, which will have samples for const hyperparameters, but we don't write an `@attribute` out for const above. If there is a const hyperparameter, would this break because the arff file would be malformed? just trying to make sure i understand.

Reply:
So when we're building the data file we first look at all the parameter types and we filter out the ones that are const. Then as we're going through the data, we only use the ones that weren't filtered out.

If we eventually refactor this into shipping the data over the network somewhere, then this would be a good thing to improve: filter out the data from the const hyperparameters earlier. But no, this won't break. In fact the integration test explicitly covers this: there are const parameters, and we check that they're excluded from HP importance results but that it doesn't otherwise break.

Reply:
Oh I see, I missed `hpsOrder` was the filtered list somehow. I understand now.